sqlgen/matcher.go

```diff
@@ -0,0 +1,118 @@
+package sqlgen
+
+import (
+	"sort"
+	"strings"
+
+	"github.com/samsarahq/thunder/internal"
+)
+
+// extractColumns extracts the columns used in a filter.
+func extractColumns(filter Filter) []string {
+	columns := make([]string, 0, len(filter))
+	for column := range filter {
+		columns = append(columns, column)
+	}
+	sort.Strings(columns)
+	return columns
+}
+
+// extractColumnValues extracts the values of a set of columns in a filter.
+func extractColumnValues(m map[string]interface{}, columns []string) []interface{} {
+	values := make([]interface{}, 0, len(columns))
+	for _, column := range columns {
+		values = append(values, m[column])
+	}
+	return values
+}
+
+// columnsKey builds a string key for a set of columns useable in a map.
+func columnsKey(columns []string) string {
+	return strings.Join(columns, ";")
+}
+
+// A matcherGroup holds a set of tuples for a given set of columns in a
+// matcher.
+type matcherGroup struct {
+	columns []string
+	tuples  map[interface{}]map[interface{}]struct{}
```

Comment:
From my read, it seems like:
- a matcher has many groups. each group corresponds to a different filter (column/value set - a "tuple")
- a group has:
  - the columns it filters on (for narrowing the matching in `match`)
    - stored in sorted order for consist array ordering in comparisons
  - a set of tuple (column/value) pairs (stored as a map{tuple value => marker})

I think it'd be nice to write out how this works in a comment.